### PR TITLE
Wildcard subscription of ddata changes

### PR DESCRIPTION
**Pull request by patriknw**
* Subscribe to keys starting with a given prefix.
* Useful together with the new expiry feature, for example.
* Also fixed a bug in the (non-released) exiry feature, gossip was broken.

TODO:
- [x] ~mima~
- [x] reference and api docs
- [x] FIXMEs

### CODE REVIEW

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 1]
```diff
@@ -2129,7 +2131,12 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
     isExpired(key, getUsedTimestamp(key))
 
   def isExpired(key: KeyId, timestamp: Timestamp): Boolean = {
-    expiryEnabled && timestamp != 0L && timestamp <= System.currentTimeMillis() - getExpiryDuration(key).toMillis
+    if (expiryEnabled && timestamp != 0L) {
+      val expiryDuration = getExpiryDuration(key)
+      expiryDuration != Duration.Zero && timestamp <= System.currentTimeMillis() - expiryDuration.toMillis
+    } else {
+      false
+    }
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala:2139]: This was a bug in the gossip exchange when expiry was configured (enabled) but exchanging other keys without expiry. getExpiryDuration is zero for those.

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/WildcardSubscribeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import akka.remote.testkit.MultiNodeConfig
+import akka.remote.testkit.MultiNodeSpec
+import akka.testkit._
+
+object WildcardSubscribeSpec extends MultiNodeConfig {
+  val first = role("first")
+  val second = role("second")
+
+  commonConfig(ConfigFactory.parseString("""
+    akka.loglevel = DEBUG
```
Comment (patriknw) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/WildcardSubscribeSpec.scala:22]: change to INFO again

**Review by johanandren — APPROVED** [review]
LGTM but additional test coverage could be good.

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/WildcardSubscribeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+        val chg2 = subscriberProbe.expectMsgType[Changed[GCounter]]
+        chg2.key should ===(KeyOtherA)
+        chg2.get(KeyOtherA).value should ===(3)
+        enterBarrier("received-2")
+      }
+    }
+
+    enterBarrier("done-3")
+  }
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/WildcardSubscribeSpec.scala:206]: added a few more in https://github.com/akka/akka/pull/31731/commits/f6675e29385c33960143cc011a98858a5d024f73

**Review by pvlugter — APPROVED** [review]
LGTM

**Pull request merged**